### ISSUE TITLE
 Implement Filter/Sort UI for Allowance Holdings Data

### ISSUE THREAD

**Issue by JorjaComer** [US-EPA-CAMD/PI-1#1009]
Original estimate: 21 
PI 6 Estimate: 8

**System Increment Definition of Ready (DoR)**
Before being accepted into a Program Increment each feature:

- [x] Meets the INVEST criteria (Independent, Negotiable, Valuable, Estimate-able, Small, Testable)
- [x] Has a clearly defined user
- [x] Has an estimate
- [x] Has acceptance criteria
- [x] Has sufficient context, such as

-User Research 
- Wireframes/UI Design
- Data Conversion (completed)
- System Requirements (in progress)

- [x] Traceable to feature, epic or PI
- [x] Dependencies are identified
- [x] Component reusability assessed

## Feature Benefit

Provide front-end user interface to access Allowance Holdings data

The following are notes for discussion. The functionality described in the user story is always the source of truth.

## Context 

[Wireframes](https://xd.adobe.com/view/129b1991-bd2b-475f-8f90-8e14d430c2ff-47c8/?fullscreen&hints=off) from 3/30/21 

## Allowance Holdings Filter/Sort UI Requirements 

Questions:

Will the site use the EPA header/footer? Will it use the go wide template? - Waiver process has begun. The header and content are fairly defined. Footer is TBD. The development will initially focus on main content, filter sort criteria, as template work continues in parallel. Pages should be responsive to screen size.

**Assumptions:**

1. Blank parameter values will return all records
2. If multiple parameters are selected in a single query (TX, general account), then only values that match all parameters will be returned (AND)
3. If multiple values are sent for a single parameter (e.g. TX, MD), then all data that matches the values is returned (OR).
4. Download limits will be implemented in future and are out of scope for initial development
5. Users must select data type and data subtype before other fields are available 
6. The site will be responsive to different desktop screen sizes (not mobile compatible) 
7. Processing the file download will be handled by the browser

**UI Parameters**

**Data Type**
Button/ Drop down 
Value:  Emissions, Allowance, Compliance, default blank, 
Required
Single select 

**Data Sub-Type** 
Drop down 
Value: Allowance Holdings,  Allowance Transactions, Account Information, default blank
Required 
Single Select 

**Vintage Year**
Text field. Allows ranges and individual years (YYYY-YYYY, YYYY, YYYY)
Format: YYYY 
Optional
Valid date range: 1995 or greater
Multi select (multiple 2019,2020, range 2017-2020, and indiv. Years 2015) 
Breadcrumb display: Vintage year (no years displayed)

**Program**	
Drop down, default blank
Values: All active programs (ARP, CSNOX, CSOSG1/2, CSSO2G1/2, TXSO2) 
Optional 
Multi select 
Out of Scope: If account is already selected, then search is limited to programs applicable to the account state  	

**Account Type**	
Check List, default blank
Values: Facility, Unit, General, Reserve, Surrender, All 
Optional
Multi select 

**Account** 

Account Name
Dynamic search/combo box
Search is a contains search on all allowance account names 
Results are displayed in Account Name (Account Number) format
Results populate in a list, parameters populates when a value is selected from the list  	(results list should stay at top of page/visible in current view, possible use pagination to keep window small)
Search is not case sensitive
Out of scope: If state is already selected, then search is limited to accounts in the selected state 	
If no results are found, the system will indicate so. “No accounts match your search.”
Multi select 

Account Number
Dynamic search/combo box
Search is a contains search on all allowance account numbers 
Results are displayed in Account Name (Account Number) format
Results populate in a list, parameters populates when a value is selected from the list 	(results list should stay at top of page/visible in current view, possible use pagination to keep window small)
Search is not case sensitive
Out of scope: If state is already selected, then search is limited to accounts in the selected state 	
If no results are found, the system will indicate so. “No accounts match your search.”
Multi select 

**Facility**

Facility Name
Dynamic search
Search is a contains search
Results are displayed in Facility Name (ORIS Code)
Results populate in a list, parameters populates when a value is selected from the list  	(results list should stay at top of page/visible in current view, possible use pagination to keep window small)
Search is not case sensitive
Out of scope: If state is already selected, then search is limited to facilities in the selected state 	
If no results are found, the system will indicate so. “No facilities match your search.”
Multi select 

ORIS Code	
Dynamic search
Search is a contains search
Results populate in a list, parameters populates when a value is selected from the list 
Results are displayed in Facility Name (ORIS Code) format
Out of scope: If state is already selected, then search is limited to facilities in the selected state 	
If no results are found, the system will indicate so. “No facilities match your search.”
Multi Select 

**Owner/ Operator** 

Owner
Dynamic search/Combo box
contains search search on all owners 
Results populate in a list, parameters populates when a value is selected from the list 	(results list should stay at top of page/visible in current view, possible use pagination to keep window small)
Search is not case sensitive
If no results are found, the system will indicate so. “No accounts match your search.”
Multi select 

Operator
Dynamic search
contains search search on all operators 
Results populate in a list, parameters populates when a value is selected from the list 	(results list should stay at top of page/visible in current view, possible use pagination to keep window small)
Search is not case sensitive
If no results are found, the system will indicate so. “No accounts match your search.”
Multi select 

State
Dynamic search
Search is a “begins with” search
Results populate in a list, parameters populates when a value is selected from the list  	
Results display the full state name (Texas, not TX)
Out of scope: If Facility/ORIS Code or Account Name/Number is already selected, then state is automatically populated with the state (or states) associated to the facility/account (or facilities/accounts) 
Multi Select

**Change Criteria**
The user will have the ability to clear one, many, or all filter criteria
If the user changes data type or sub data type, then:

The filter criteria displayed updates accordingly
No previously entered filter criteria is retained

If a user changes data type or sub data type, then filter criteria is cleared

If the user chooses to clear one or more filter criteria before preview is selected, then the system clears the criteria from the filter and updates any remaining dependent criteria accordingly.

If the user chooses to clear one or more filter criteria after preview is selected, then the system clears the criteria from the filter AND download is disabled until data is previewed.

**Data Preview**
Trigger: A user has entered, at a minimum, the required parameters, and selected the option to preview data.
For each record that matches the criteria, the system will display in a tabular format:

ACCOUNT_NUMBER
ACCOUNT_NAME
ORISPL_CODE
PRG_CODE
VINTAGE_YEAR
TOTAL_BLOCK
START_BLOCK
END_BLOCK
STATE
EPA_REGION (2 digit number)
OWN_DISPLAY
ACCOUNT_TYPE

**Data Labels-** 
Refer to the CAMPD_ DataLabels_AllowanceCompliance file [here ](https://usepa.sharepoint.com/:x:/r/sites/CAMDCVPTeam/_layouts/15/Doc.aspx?sourcedoc=%7B7A7D0213-1149-4060-B470-19DE87101DF1%7D&file=CAMPD_DataLabels_AllowanceCompliance_3_18_2021.xlsx&action=default&mobileredirect=true)

Results will display record count
Results are limited to 100 records
Results in preview will have a scroll functionality to show all columns/rows
NOTE: NEED TO ID COLUMNS FOR HOVER TEXT AND DISPLAY TEXT
OUT OF SCOPE: Results may be sorted in alphabetical ascending and descending order
No additional filtering: CAN THEY BE FILTERED? IF SO BY WHAT CRITERIA?
OUT OF SCOPE: Results are paginated

**File Selection**
Trigger: The user has at least one record that matches the search criteria, and has previewed the results
The user has the option to select the file output in JSON or CSV format. CSV is the default file type 
API will provide the data in the requested format based on user selection in UI

**Download**
Trigger: The user has selected a file output type
When the user selects the option to download, the system creates a file in the format specified  containing all records that match the parameters and downloads the file 
After a user has chosen to download data, then the system will display the preview of the downloaded results.
Error Conditions
**API Timeout**
Default API timeout = TBD
When default time out is reached, API will generate error condition. Display message in UI. SAMPLE TEXT: The data you requested is too large to download. Refine your criteria to limit your results, or visit the Datasets page <insert link to data sets page>. 

**Bad request**

**Internal server error/network outage** 

New FC = 3 (4)
Existing FC = 1 (3)
Preview = 2 
Filter Tags = 2
Download = 3

**Comment by tlageman-epa** [comment]
@JanellC I think we are missing a ticket for the Vintage year filter. Not seeing it under this epic. Thanks!

**Comment by JanellC** [comment]
@tlageman-epa Thank you for pointing this out. For some reason AH: Filter Custom Data Download By Vintage Year #1102 was closed. I reopened the issue and the feature should reflect the change.

**Comment by tlageman-epa** [comment]
@JanellC Same as your other question. Buyer/seller accounts are not applicable here. This data is holdings data for every allowance account for all programs. This is not transaction data.

**Comment by JanellC** [comment]
Completed core functionality. Allowance holdings functions on the custom data download with end to end functionality. Removed tickets #1999 , #2110, and #2089 and from this epic. The removed issues remain on the PI 6 backlog
